File: DIP/Spoilers/MarsRoverKata/Rover.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace MarsRoverKata
{
    public class Rover : CombatMoveable
    {
        public Rover(Mars mars)
            : this(mars, mars.CenterOfThePlanet)
        { }

        public Rover(Mars mars, Point landingPoint)
            : base(mars, landingPoint)
        {
            Projectiles = new List<Projectile>
            {
                new Missile(Mars),
                new Missile(Mars),
                new Missile(Mars),
                new Mortar(Mars),
                new Mortar(Mars),
                new Mortar(Mars)
            };
            LandOnMars(landingPoint);
        }

        private void LandOnMars(Point landingPoint)
        {
            if (!Mars.IsValidPosition(landingPoint))
                throw new CrashException("Doh! We tried to land on something other than the planet and the rover was destroyed!!!");
            Facing = Direction.North;
        }
    }
}

File: LSP+ISP/Java_Final/MarsRoverKata/src/main/java/MarsRoverKata/Rover.java
package MarsRoverKata;

import com.google.common.base.Predicate;
import com.google.common.collect.Collections2;
import com.google.common.collect.Iterables;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Rover extends Movable {
    private List<Projectile> projectiles;

    public Rover(Mars mars) throws CrashException {
        this(mars, mars.getCenterOfThePlanet());
    }

    public Rover(final Mars mars, final Point landingPoint) throws CrashException {
        super(mars);
        projectiles = new ArrayList<Projectile>() {
            {
                add(new Missile(mars));
                add(new Missile(mars));
                add(new Missile(mars));
                add(new Mortar(mars));
                add(new Mortar(mars));
                add(new Mortar(mars));
            }
        };
        landOnMars(landingPoint);
    }

    private void landOnMars(Point landingPoint) throws CrashException {
        if (!getMars().isValidPosition(landingPoint)) {
            throw new CrashException("Doh! We tried to land on something other than the planet and the rover was destroyed!!!");
        }
        setLocation(landingPoint);
        setFacing(Direction.North);
    }

    public boolean fireProjectile(final Class<? extends Projectile> type) {
        Projectile projectileToFire = Iterables.get(Collections2.filter(projectiles, new Predicate<Projectile>() {
            @Override
            public boolean apply(Projectile input) {
                return (input.getClass().isAssignableFrom(type));
            }
        }), 1, null);
        if (projectileToFire == null) {
            return false;
        }
        projectileToFire.launch(getFacing(), getLocation());
        projectiles.remove(projectileToFire);
        return true;
    }
}

